ruby/ql/lib/codeql/ruby/security/UnsafeHtmlConstructionQuery.qll

```diff
@@ -0,0 +1,49 @@
+/**
+ * Provides a taint-tracking configuration for reasoning about HTML
+ * constructed from library input vulnerabilities.
+ *
+ * Note, for performance reasons: only import this file if `Configuration` is needed,
+ * otherwise `UnsafeHtmlConstructionCustomizations` should be imported instead.
+ */
+
+import codeql.ruby.DataFlow
+import UnsafeHtmlConstructionCustomizations::UnsafeHtmlConstruction
+private import codeql.ruby.TaintTracking
+private import codeql.ruby.dataflow.BarrierGuards
+
+/**
+ * A taint-tracking configuration for detecting unsafe HTML construction.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnsafeShellCommandConstruction" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node instanceof StringConstCompareBarrier or
+    node instanceof StringConstArrayInclusionCallBarrier
+  }
+
+  // override to require the path doesn't have unmatched return steps
+  override DataFlow::FlowFeature getAFeature() {
+    result instanceof DataFlow::FeatureHasSourceCallContext
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
```

Comment:
Are these additional steps necessary? I would have assumed that they would be covered by the flow summaries in [Array.qll](https://github.com/github/codeql/blob/main/ruby/ql/lib/codeql/ruby/frameworks/core/Array.qll).

Reply:
They are, because steps in `Array.qll` will only propagate taint from a store to the object to a read from the object.  
This taints the object itself.  

But that is mostly relevant when an array is a relevant sink, which is the case for code-construction (where I copied the code from, `.join(" ")` is used to concat shell arguments).  
But I didn't use that sink here, so the taint-step is not really relevant.  
I've deleted it.